### PR TITLE
Build system cleanup

### PR DESCRIPTION
**Pull request by SamB**
I'm trying to clean up some things about the build system, among them the near-duplicate -proto projects (#712), but I don't actually want my changes pulled yet; I'm only creating this pull request in order to [ab]use the CI infrastructure.

Edit: It seems like this might be a good time to merge what I've done so far, as you seem to like it and it might take me a while to script the generation of the -proto projects.

### CODE REVIEW

**Comment by SamB** [comment]
By the way, I've been using this little shell script to compare the -proto projects with the normal versions;
```
#!/bin/sh
for i in FSharp.Build FSharp.Compiler Fsc; do
    git --no-pager diff --no-index src/fsharp/{$i/$i,$i-proto/$i-proto}.fsproj;
done
```
(Don't forget `chmod +x`; I was a little surprised that this was needed on Windows since I thought most files had execute permissions here, but for some reason I had to do that before bash autocompletion worked.)

Also, my laptop power cord is busted, plus I'm going to be visiting with a friend I haven't seen in a while, so I may be out of touch for at least the next couple of days.

**Comment by KevinRansom** [comment]
Wow!!! thanks for diving into the build, you are surely a brave person.  We need all of the help we can get with making it a bit less opaque.

**Comment by SamB** [comment]
Oh dear, some sort of phase-crossing seems to have crept in somehow :-(

File: src/fsharp/FSharp.Compiler/FSharp.Compiler.fsproj [hunk 1]
```diff
@@ -456,6 +462,21 @@
     <Compile Include="..\fsc.fs">
       <Link>Driver\fsc.fs</Link>
     </Compile>
+    <!--
```
Comment (dsyme) [inline src/fsharp/FSharp.Compiler/FSharp.Compiler.fsproj]: Please remove this comment - it's only understandable in the context of the specific "no duplication" aims being discussed in this PR. thanks

**Comment by dsyme** [comment]
I'm generally ok with the PR as it is at the time of writing this comment - it achieves good textual alignment between Proto and non-Proto.  

However it's probably enough for now?  For the reasons in [this comment](https://github.com/Microsoft/visualfsharp/issues/712#issuecomment-154218909).

**Comment by SamB** [comment]
@dsyme: Hmm. That would *also* save me from having to figure out WTH the `Configuration` isn't getting set the way I'd expect in the final projects, which I was thinking might require me to add more diagnostics to MSBuild.

But if doing away with the -proto projects isn't an option (and turning them into thin wrappers around the normal projects is also out), then I guess I'd want to generate them as part of the build, so they're forced to be in sync with the main ones.  (And that it might make the logistics easier to not do it using F# code; C# would likely be simplest, or possibly VB?)

**Comment by enricosada** [comment]
> then I guess I'd want to generate them as part of the build, so they're forced to be in sync with the main ones

@SamB dont add anoter pre build step pls. ideal workflow is:

- clone repo
- open solution
- run

there are already lots of prebuild step requirements, and build infrastructure is already weak (.bat files, multiple steps)

Maybe add a command ( .fsx? ) to generate it, so we can use it and **commit** generated/updated -proto proj to repo.
if we need to check sync, we can add a test to check if generated output is the same of actual file in the repo
File list not change alot, no need to recreate the -proto proj before each build

**Comment by dsyme** [comment]
@SamB   Yes, I think the ideal for now is to still have the two project files, just have them much better aligned.

**Comment by enricosada** [comment]
btw @SamB thx for your work, the infrastructure/build system need some love

**Comment by SamB** [comment]
@enricosada Hmm, yes, a manually-run script with an assert somewhere to verify freshness would certainly work.

And I've looked at trying to reduce the reliance on batch scripting in the build; in fact, I have an MSBuild `.proj` stashed somewhere that's meant to distill each batch of `msbuild` invocations in [section 3 of DEVGUIDE](https://github.com/Microsoft/visualfsharp/blob/master/DEVGUIDE.md#3-full-steps-before-running-tests) down to just one invocation, and it seemed to work for me.  It sounds like you might appreciate it if I would polish that off and submit it sooner rather than later?

**Comment by enricosada** [comment]
@SamB yes, that part is really annoying, i usually do appveyor.bat and a custom script but is not a solution for new contributors.
Without changing the build system, helper commands (like your msbuild proj to run bat) may help.
Or another bat file, sigh

**Comment by SamB** [comment]
@enricosada: Actually, it doesn't run bat, it does recursive MSBuild on all the relevant projects, including FSharp.Core and its tests for all of the different target frameworks that DEVGUIDE lists. One of the nice things about this is that it lets MSBuild summarize **all** of the errors at the end.

**Comment by enricosada** [comment]
a lot better if summarize all errors :smile: ( + stop at first failing project if possible )

**Comment by SamB** [comment]
@enricosada: So, kind of like this?

**Comment by SamB** [comment]
@dsyme So is there anything else that I need to do before this bit is merged?

**Comment by KevinRansom** [comment]
The requirements for cross platform will require a different approach.

Closing for now.

**Pull request closed without merging**